### PR TITLE
Add support for PAW3204 Optical Sensor

### PR DESCRIPTION
**Pull request by drashna**
## Description

Adds support for the PAW3204 low powered optical sensor.

This is basic support, as the device supports multiple power modes and more, but aren't actually implemented here.

Uses bitbanged communication

## Types of Changes

- [x] Core
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: drivers/sensors/paw3204.c [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    return byte;
+}
+
+void paw3204_serial_write(uint8_t data) {
+    datatogglestate = readPin(PAW3204_SDIO_PIN);
+    if (datatogglestate == 1) {
+        writePinLow(PAW3204_SDIO_PIN);
+    } else {
+        writePinLow(PAW3204_SDIO_PIN);
+    }
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:83]: Both branches write the pin low, is this correct?

File: drivers/sensors/paw3204.c [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    wait_us(4);
+}
+
+int8_t convert_twoscomp(uint8_t data) {
+    if ((data & 0x80) == 0x80)
+        return -128 + (data & 0x7F);
+    else
+        return data;
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:103]: Parenthesis should be added to the branches.

File: drivers/sensors/paw3204.c [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    }
+    setPinOutput(PAW3204_SDIO_PIN);
+
+    for (int8_t b = 7; b >= 0; b--) {
+        writePinLow(PAW3204_SCLK_PIN);
+        if (data & (1 << b))
+            writePinHigh(PAW3204_SDIO_PIN);
+        else
+            writePinLow(PAW3204_SDIO_PIN);
+        // wait_us(2);
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:92]:
```suggestion
```

If it is not needed at all?

**Review by KarlK90 — COMMENTED** [review]
Nice! Do we have any prove that the driver works? As I don't have the sensor myself this would be good.

**Comment by drashna** [comment]
> Nice! Do we have any prove that the driver works? As I don't have the sensor myself this would be good.

Unfortunately, no.  The code is taken from a fork. And unfortunately, I don't have actual hardware to test on.

File: drivers/sensors/paw3204.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return data;
+    }
+}
+
+report_paw3204_t paw3204_read(void) {
+    report_paw3204_t data = {0};
+
+    data.isMotion = paw3204_read_reg(REG_STAT) & (1 << 7); // check for motion only (bit 7 in field)
+    data.x        = convert_twoscomp(paw3204_read_reg(REG_X));
+    data.y        = convert_twoscomp(paw3204_read_reg(REG_Y));
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:110]:
```suggestion
    data.x        = (int8_t)paw3204_read_reg(REG_X);
    data.y        = (int8_t)paw3204_read_reg(REG_Y);
```

File: drivers/sensors/paw3204.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ */
+
+// https://github.com/shinoaliceKabocha/choco60_track/tree/master/keymaps/default
+
+#include "paw3204.h"
+#include "wait.h"
+#include "debug.h"
+#include "gpio.h"
+
+uint8_t datatogglestate;
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:24]:
```suggestion
```
This variable is only written to once and never used.

File: drivers/sensors/paw3204.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+uint8_t paw3204_read_reg(uint8_t reg_addr) {
+    uint8_t byte = 0;
+
+    paw3204_serial_write(reg_addr);
+    wait_us(5);
+    byte = paw3204_serial_read();
+
+    return byte;
+}
```
Comment (KarlK90) [inline drivers/sensors/paw3204.c:128]:
```suggestion
uint8_t paw3204_read_reg(uint8_t reg_addr) {
    paw3204_serial_write(reg_addr);
    wait_us(5);
    return paw3204_serial_read();
}
```

**Review by KarlK90 — CHANGES_REQUESTED** [review]
More thorough review of the code. I'm beginning to think that we should get somebody with a PAW3204 to verify that this driver actually works. (Maybe the original author?)

**Comment by drashna** [comment]
> More thorough review of the code. I'm beginning to think that we should get somebody with a PAW3204 to verify that this driver works an intended. ~(Maybe the original author?)~ Pinged Gompa on Discord.

Yeah, ideally, we should. 

That said, @GEIGEIGEIST (?) was designing a PCB around this sensor, which is what kicked this off for me.

**Comment by KarlK90** [comment]
Gompa is away from their computer for a week, but will report back afterwards.

**Comment by GEIGEIGEIST** [comment]
I'm back at my computer/my keyboard on Monday and will report back then.

**Comment by freznel10** [comment]
I tested on a blackpill and can confirm that it works.

**Comment by GEIGEIGEIST** [comment]
I tested it on my PCB, running on a RP2040 Pro Micro and it works too. Thank you!

**Review by KarlK90 — APPROVED** [review]
One last round of review comments (hope you don't feel nitpicked) and then this is good to go from my side.

**Comment by drashna** [comment]
> One last round of review comments (hope you don't feel nitpicked) and then this is good to go from my side.

Nope, not at all.  Helps me get into better practice :)

And applied.

**Comment by KarlK90** [comment]
> > One last round of review comments (hope you don't feel nitpicked) and then this is good to go from my side.
> 
> Nope, not at all. Helps me get into better practice :)
> 
> And applied.

Good to hear! Feel free to merge.

**Pull request merged**